fuel-p2p/src/discovery.rs

```diff
@@ -0,0 +1,507 @@
+        self.kademlia.inject_expired_listen_addr(id, addr);
+    }
+
+    fn inject_dial_failure(
+        &mut self,
+        peer_id: Option<PeerId>,
+        handler: Self::ProtocolsHandler,
+        err: &DialError,
+    ) {
+        self.kademlia.inject_dial_failure(peer_id, handler, err)
+    }
+
+    fn inject_new_listen_addr(&mut self, id: ListenerId, addr: &Multiaddr) {
+        self.kademlia.inject_new_listen_addr(id, addr)
+    }
+
+    fn inject_listener_error(&mut self, id: ListenerId, err: &(dyn std::error::Error + 'static)) {
+        self.kademlia.inject_listener_error(id, err)
+    }
+
+    fn inject_listener_closed(&mut self, id: ListenerId, reason: Result<(), &io::Error>) {
+        self.kademlia.inject_listener_closed(id, reason)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{DiscoveryBehaviour, DiscoveryConfig};
+    use crate::discovery::DiscoveryEvent;
+    use futures::{future::poll_fn, StreamExt};
+    use libp2p::{
+        core, identity::Keypair, multiaddr::Protocol, noise, swarm::SwarmEvent, yamux, Multiaddr,
+        PeerId, Swarm, Transport,
+    };
+    use std::{
+        collections::{HashSet, VecDeque},
+        task::Poll,
+    };
+
+    /// helper function for building Discovery Behaviour for testing
+    fn build_fuel_discovery(
+        predefined_peers: Vec<(PeerId, Multiaddr)>,
+    ) -> (Swarm<DiscoveryBehaviour>, Multiaddr, PeerId) {
+        let keypair = Keypair::generate_secp256k1();
+        let public_key = keypair.public();
+
+        let noise_keys = noise::Keypair::<noise::X25519Spec>::new()
+            .into_authentic(&keypair)
+            .unwrap();
+
+        let transport = core::transport::MemoryTransport
+            .upgrade(core::upgrade::Version::V1)
+            .authenticate(noise::NoiseConfig::xx(noise_keys).into_authenticated())
+            .multiplex(yamux::YamuxConfig::default())
+            .boxed();
+
+        let behaviour = {
+            let mut config =
+                DiscoveryConfig::new(keypair.public().to_peer_id(), "test_network".into());
+            config
+                .discovery_limit(50)
+                .with_predefined_nodes(predefined_peers)
+                .enable_random_walk(true);
+
+            config.finish()
+        };
+
+        let listen_addr: Multiaddr = Protocol::Memory(rand::random::<u64>()).into();
+        let mut swarm = Swarm::new(transport, behaviour, keypair.public().to_peer_id());
+
+        swarm
+            .listen_on(listen_addr.clone())
+            .expect("swarm should start listening");
+
+        (swarm, listen_addr, PeerId::from_public_key(&public_key))
+    }
+
+    // builds 25 discovery swarms,
+    // initially, only connects first_swarm to the rest of the swarms
+    // after that each swarm uses kademlia to discover other swarms
+    // test completes after all swarms have connected to each other
+    #[tokio::test]
+    async fn discovery_works() {
+        env_logger::init();
```

Comment:
Is `env_logger` required for libp2p? Or can we use `tracing` instead?

Reply:
definitely not required, it was first thing that came to my mind when testing

Reply:
since it's just a dev-dep that's fine. Although we prefer to use tracing in normal code.